common/src/java/org/apache/hadoop/hive/common/FileUtils.java

```diff
@@ -408,10 +408,23 @@ public static FileStatus getPathOrParentThatExists(FileSystem fs, Path path) thr
     return getPathOrParentThatExists(fs, parentPath);
   }
 
-  public static void checkFileAccessWithImpersonation(final FileSystem fs, final FileStatus stat,
-      final FsAction action, final String user)
-      throws IOException, AccessControlException, InterruptedException, Exception {
-    checkFileAccessWithImpersonation(fs, stat, action, user, null);
+  public static void checkFileAccessWithImpersonation(final FileSystem fs, final FileStatus stat, final FsAction action,
+      final String user) throws IOException, AccessControlException, InterruptedException, Exception {
+    UserGroupInformation ugi = Utils.getUGI();
+    String currentUser = ugi.getShortUserName();
+    UserGroupInformation proxyUser = null;
+    FileSystem fsAsUser = null;
+    try {
+      if (user != null && !user.equals(currentUser)) {
+        proxyUser = UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser());
+        fsAsUser = FileUtils.getFsAsUser(fs, proxyUser);
+      }
+      checkFileAccessWithImpersonation(fs, stat, action, user, null, fsAsUser);
```

Comment:
instead of fsAsUser, pass currentUser and do proxy user creation in checkFileAccessWithImpersonation method.

Reply:
Moving the FIleSystem.getFsAsUser() outside checkFileAccessWithImpersonation is the optimization. Just initializing once for the fs and proxyUser.

Reply:
Is the optimisation preventing doing multiple times FileSystem.get() for same user and FileSystem to prevent multiple `` fs.initialize(uri, conf);`` calls?
In that case even if we call FileSystem.get() multiple times it would do init only once, from the next time it will return the same FileSystem instance from the cache.

It is there in the Javadoc as well
```
   *   If the there is a cached FS instance matching the same URI, it will
   *   be returned.
```

Reply:
Yes I looked at that. But right now we can see that the FIleSystem.get() (either from cache or new init object) takes few ms for each file accumulating few seconds if the number of files is too many. We would like to completely get rid of the FS.get() calls inside a recursive path -- unless it is necessary